src/macaron/vsa/vsa.py

```diff
@@ -0,0 +1,251 @@
+    verification_result : VerificationResult
+        The verification result of the subject.
+
+    Returns
+    -------
+    VsaStatement
+        A Statement layer of the VSA.
+    """
+    return VsaStatement(
+        _type="https://in-toto.io/Statement/v1",
+        subject=[
+            {
+                "uri": subject_purl,
+            }
+        ],
+        predicateType="https://slsa.dev/verification_summary/v1",
+        predicate=VsaPredicate(
+            verifier=Verifier(
+                id="https://github.com/oracle/macaron",
+                version={
+                    "macaron": importlib_metadata.version("macaron"),
+                },
+            ),
+            timeVerified=datetime.datetime.now(tz=datetime.UTC).isoformat(),
+            resourceUri=subject_purl,
+            policy={
+                "content": policy_content,
+            },
+            verificationResult=verification_result,
+            verifiedLevels=[],
+        ),
+    )
+
+
+def get_subject_verification_result(policy_result: dict) -> tuple[str, VerificationResult] | None:
+    """Get the PURL (string) and verification result of the single software component the policy applies to.
+
+    This is currently done by reading the facts of two relations:
+    ``component_violates_policy``, and ``component_satisfies_policy``
+    from the result of the policy engine.
+
+    We define two PURLs to be different if the two PURL strings are different.
+
+    The result of this function depends on the policy engine result:
+
+    - If there exist multiple different PURLs, this function returns ``None``.
+    - If there exist multiple occurrences of the same PURL and it is the only unique
+      PURL in the policy engine result, this function returns the latest occurrence,
+      which is the PURL that goes with the highest component ID, taking advantage of
+      component IDs being auto-incremented.
+    - If there is no PURL in the result, i.e. the policy applies to no software component
+      in the database, this function also returns ``None``.
+
+    Parameters
+    ----------
+    policy_result : dict
+        The result of the policy engine, including two relations:
+        ``component_violates_policy``, and ``component_satisfies_policy``.
+
+    Returns
+    -------
+    tuple[str, VerificationResult] | None
+        A pair of PURL and verification result of the only software component that
+        the policy applies to, or ``None`` according to the aforementioned conditions.
+    """
+    component_violates_policy_facts = policy_result.get("component_violates_policy", [])
+    component_satisfies_policy_facts = policy_result.get("component_satisfies_policy", [])
+
+    # key: PURL; value: result with the highest component id
+    component_results: dict[str, tuple[int, VerificationResult]] = {}
+
+    for component_id_string, purl, _ in component_violates_policy_facts:
+        component_id = int(component_id_string)
```

Comment:
To catch unexpected `component.id`s, e.g., if it is changed to UUID but this code is not updated, it would be good to check for `ValueError` and log an error.

Reply:
This function currently relies on the assumption that the component id column is auto-incremented to pick the latest analysis results. I don't think UUID is applicable here since there is no ordering on UUID, and if we switch to UUID then the implementation will need to change and this function will most likely become irrelevant.

Reply:
Yes that's right. I just want to be precautious and log errors if the assumption made in this implementation is violated.

Reply:
See 60ca8cf3970350aeb6df1a6b260a3d4189571969.